service/controller/chart/v1/resource/releasemigration/resource.go

```diff
@@ -53,3 +70,103 @@ func New(config Config) (*Resource, error) {
 func (r *Resource) Name() string {
 	return Name
 }
+
+func (r *Resource) findHelmV2Releases(ctx context.Context) ([]string, error) {
+	lo := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", "OWNER", "TILLER"),
+	}
+
+	// Check whether helm 2 release configMaps still exist.
+	cms, err := r.k8sClient.CoreV1().ConfigMaps(r.tillerNamespace).List(lo)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	hadReleases := map[string]bool{}
+	for _, cm := range cms.Items {
+		name, _ := cm.GetLabels()["NAME"]
+		if _, ok := hadReleases[name]; !ok {
+			hadReleases[name] = true
+		}
+	}
+
+	releases := make([]string, 0, len(hadReleases))
+	for k := range hadReleases {
+		releases = append(releases, k)
+	}
+
+	return releases, nil
+}
+
+func (r *Resource) ensureReleasesMigrated(ctx context.Context) error {
+	// 0. Found all dangling helm release v2
+	releases, err := r.findHelmV2Releases(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// 1. Install helm-2to3-migration app
+	{
+		var tarballPath string
+		{
+			tarballURL, err := appcatalog.GetLatestChart(ctx, key.DefaultCatalogStorageURL(), "helm-2to3-migration")
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			tarballPath, err = r.helmClient.PullChartTarball(ctx, tarballURL)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			defer func() {
+				fs := afero.NewOsFs()
+				err := fs.Remove(tarballPath)
+				if err != nil {
+					r.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("deletion of %#q failed", tarballPath), "stack", fmt.Sprintf("%#v", err))
+				}
+			}()
+
+			opts := helmclient.InstallOptions{
+				ReleaseName: "helm-2to3-migration",
+			}
+
+			values := map[string]interface{}{
+				"releases": releases,
+				"tiller": map[string]string{
+					"namespace": r.tillerNamespace,
+				},
+			}
+
+			err = r.helmClient.InstallReleaseFromTarball(ctx, tarballPath, r.tillerNamespace, values, opts)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	// 2. Wait until all helm v2 release are deleted
+	operation := func() error {
+		releases, err := r.findHelmV2Releases(ctx)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(releases) > 0 {
+			return microerror.Maskf(releasesNotDeletedError, "helm v2 releases not deleted: %#q", releases)
+		}
+
+		return nil
+	}
+
+	notify := func(err error, t time.Duration) {
+		r.logger.Log("level", "debug", "message", "failed to deleted all helm v2 releases")
+	}
+
+	b := backoff.NewConstant(backoff.ShortMaxWait, backoff.ShortMaxInterval)
```

Comment:
I like using the actual time intervals in backoffs. I always forget the constant values so for me its more readable.

The migration is an unusual situation but I have a worry about using a backoff for this. Could we cancel the reconciliation and check again in the next loop?

Reply:
Mostly in the next loop, it will reconcile on the next chart CRs, and would try to gather helm v2 release and would attempt to install `helm-2to3-migration` app. 
I would need to put an error handling guessing on this situation, but does it worth than this backoff? I would like to keep on this backoff for this reason. 

And What do you think having 3 seconds for interval and 30 seconds for the maximum wait?

Reply:
OK cool I'm fine with keeping the backoff. 👍 I think those settings are good. 

But the migration may go wrong. Maybe we should just check if the migration app release exists?

If so we can cancel the resource. We will get alerted because the migration did not complete. WDYT?

Reply:
> We will get alerted because the migration did not complete. WDYT? 

If we canceled the resource, we get paged? Don't we need to put an unhandled error?

Reply:
No we need to add a new metric for un-migrated releases. We can alert based on that.

https://github.com/giantswarm/giantswarm/blob/master/areas/managed-services/specs/helm-3-upgrade.md#non-migrated-helm-releases-metric

Reply:
Thanks, May I can start from the next PR?

Reply:
Yes definitely. Better in a separate PR.